csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,315 @@
+  //! This lets us find starting IterDomains for scheduling concrete's loop
+  //! domain.
+  //! TODO: this should probably use a new more general utility that resembles
+  //! StmtSort::getExprsTo in val_graph_visitor.h
+  void findScheduledValGroups() {
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_tensor_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups_.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (const ValGroup& vg_inp : graph_->inputGroups(eg)) {
+          vg_stack.push(vg_inp);
+        }
+      }
+    }
+  }
+
+  //! Look at all ValGroups that are producers of abstract_tensor_.domain and
+  //! map any of those containing loop IterDomains in tv to those loop
+  //! IterDomains. These provide the starting points for scheduling.
+  GroupIdMap mapScheduledGroupsToLoopIterDomains(TensorView* tv) const {
+    GroupIdMap computed_ids;
+
+    // Look up each loop IterDomain in tv and assert that it is in
+    // scheduled_val_groups_, then map it.
+    for (IterDomain* loop_id : tv->getLoopDomain()) {
+      ValGroup vg = graph_->toGroup(loop_id);
+      // NOTE: this check guarantees that we will only ever _append_ transforms
+      // to the loop domain of tv. If instead we searched for producer IDs in
+      // the root->loop path of tv, then we could wind up silently erasing
+      // some pre-existing transforms.
+      NVF_CHECK(
+          scheduled_val_groups_.count(vg) > 0,
+          "Found loop IterDomain ",
+          loop_id->toString(),
+          " that is not described by abstract tensor.");
+      computed_ids.emplace(vg, loop_id);
+    }
+
+    return computed_ids;
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id, GroupIdMap& computed_ids) {
+    const ValGroup& g = abstractIdToValGroup(abs_id);
+
+    // Uncomputable ValGroups are those that we cannot compute from tv's root.
+    // When we detect that any inputs of an ExprGroup are in these groups, we
+    // know that we cannot compute that ExprGroup so we should skip it when
+    // scheduling tv.
+    //
+    // For example, suppose we had
+    //   tv:
+    //     logical/loop: iS0{i0} iS1{i1} iS2{i2}
+    //
+    //   graph:
+    //     ValGroup 0: iS0
+    //     ValGroup 1: iS1
+    //     ValGroup 2: iS2 iS6
+    //     ValGroup 3: iS3
+    //     ValGroup 4: iS4
+    //     ValGroup 5: iS5
+    //     ValGroup 6: iS7
+    //     ValGroup 7: iS8
+    //     ExprGroup 0:
+    //       iS3 = merge(iS0, iS1)
+    //     ExprGroup 1:
+    //       iS4, iS5 = split(iS3, 16)
+    //     ExprGroup 2:
+    //       iS8 = merge(iS6, iS7)
+    //
+    //   abstract_tensor_.domain: ValGroup 4, ValGroup 5, ValGroup 7
+    //
+    //      VG0   VG1
```

Comment:
I think it'd be helpful to emphasize that this graph represents the transformations embedded in the abstract tensor and that only `VG0`, `VG1` and `VG2` are the ones that already for the tensor. When I first saw it, I mistakenly thought the left tree represented the transformation of the tensor, which isn't correct.

Reply:
I expanded on the comment and added another example showing a ValGroup that is safe to ignore since it has no computable producer groups, as well as one that will cause an error since it has multiple producer groups and only some of them are computable.